src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs

```diff
@@ -0,0 +1,192 @@
+                    var liveDiagnostics = ConvertToLiveDiagnostics(lookup, descriptors, poolObject.Object);
+
+                    // REVIEW: for now, we are putting build error in document state rather than creating its own state.
+                    //         reason is so that live error can take over it as soon as possible
+                    //         this also means there can be slight race where it will clean up eventually. if live analysis runs for syntax but didn't run
+                    //         for document yet, then we can have duplicated entries in the error list until live analysis catch.
+                    await PersistAndReportAsync(stateSet, StateType.Document, document, textVersion, semanticVersion, liveDiagnostics).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private bool PreferBuildErrors(Workspace workspace)
+        {
+            return workspace.Options.GetOption(InternalDiagnosticsOptions.BuildErrorIsTheGod) || workspace.Options.GetOption(InternalDiagnosticsOptions.BuildErrorWinLiveError);
+        }
+
+        private async Task PersistAndReportAsync(
+            StateSet stateSet, StateType stateType, Document document, VersionStamp textVersion, VersionStamp semanticVersion, ImmutableArray<DiagnosticData> diagnostics)
+        {
+            var state = stateSet.GetState(stateType);
+            var existingDiagnostics = await state.TryGetExistingDataAsync(document, CancellationToken.None).ConfigureAwait(false);
+
+            var mergedDiagnostics = MergeDiagnostics(diagnostics, GetExistingDiagnostics(existingDiagnostics));
+            await state.PersistAsync(document, new AnalysisData(textVersion, semanticVersion, mergedDiagnostics), CancellationToken.None).ConfigureAwait(false);
+            RaiseDiagnosticsUpdated(stateType, document.Id, stateSet.Analyzer, new SolutionArgument(document), mergedDiagnostics);
+        }
+
+        private static ImmutableArray<DiagnosticData> GetExistingDiagnostics(AnalysisData analysisData)
+        {
+            if (analysisData == null)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            return analysisData.Items;
+        }
+
+        private ImmutableArray<DiagnosticData> MergeDiagnostics(ImmutableArray<DiagnosticData> liveDiagnostics, ImmutableArray<DiagnosticData> existingDiagnostics)
+        {
+            ImmutableArray<DiagnosticData>.Builder builder = null;
+
+            if (liveDiagnostics.Length > 0)
+            {
+                builder = ImmutableArray.CreateBuilder<DiagnosticData>();
+                builder.AddRange(liveDiagnostics);
+            }
+
+            if (existingDiagnostics.Length > 0)
+            {
+                builder = builder ?? ImmutableArray.CreateBuilder<DiagnosticData>();
+                builder.AddRange(existingDiagnostics.Where(d => d.Severity == DiagnosticSeverity.Hidden));
```

Comment:
I am not sure whether csc or vbc allow that, but csc,vbc never output hidden one.